File: generate.py
from typing import Set, Tuple, List

from dataclasses import dataclass
from tqdm import tqdm
import numpy as np

from util import (
    convert_matrix_to_letters,
    all_matrix_slices,
    all_matrix_slices_at,
)
from datatypes import GridState
from trie import WordTrie


def score_word(
    coverage: np.ndarray,
    open_slots: np.ndarray,
    word_indices: Tuple[List[int], List[int]],
    word_rank: float,
    random_value: float,
) -> float:
    word_intersection = np.sum(coverage[word_indices])
    fills_open_slots = np.sum(open_slots[word_indices])

    # If placing the word would not change our board, then we don't care
    if fills_open_slots == 0:
        return 0

    # Now, we want to prioritize intersecting with existing words,
    # and focus on words which are more frequent, if possible
    word_score = (
        1.5 * word_intersection
        + 0.5 * fills_open_slots
        + 2 * word_rank
        + 0.2 * random_value
    )
    return word_score


@dataclass
class WordSearch:
    word_grid: np.ndarray
    words: Set[str]


def grid_coverage(grid: np.ndarray, trie: WordTrie) -> np.ndarray:
    coverage_grid = np.zeros_like(grid, dtype=bool)

    for indices, slice in all_matrix_slices(grid):
        slice_coverage = trie.coverage(slice)
        coverage_grid[indices] = coverage_grid[indices] | slice_coverage

    return coverage_grid


def grid_words(grid: np.ndarray, trie: WordTrie) -> Set[str]:
    words = set()

    for _, slice in all_matrix_slices(grid):
        words.update(trie.words(slice))

    return words


def at_least_one_option(
    word_search_grid: np.ndarray,
    blocked: np.ndarray,
    trie: WordTrie,
    x: int,
    y: int,
) -> bool:
    for slice_indices, slice, index_in_slice in all_matrix_slices_at(
        word_search_grid,
        x,
        y,
    ):
        for _ in trie.get_valid_words(
            slice=slice, blocked=blocked[slice_indices], target_index=index_in_slice
        ):
            return True
    return False


def at_least_one_option_per_non_blocked_spot(
    word_search_grid: np.ndarray,
    blocked: np.ndarray,
    coverage: np.ndarray,
    trie: WordTrie,
) -> bool:
    for x, y in zip(*np.where(~(coverage | blocked))):
        if not at_least_one_option(
            word_search_grid=word_search_grid,
            blocked=blocked,
            trie=trie,
            x=x,
            y=y,
        ):
            return False
    return True


def fill_grid_backtracking(
    word_grid: np.ndarray,
    coverage: np.ndarray,
    trie: WordTrie,
    progress_bar: tqdm = None,
    max_checked_locations: int = 50,
) -> Tuple[np.ndarray, np.ndarray] | None:
    if progress_bar:
        num_filled = np.sum(word_grid != GridState.OPEN)
        progress_bar.n = num_filled
        progress_bar.refresh()

    uncovered_and_not_blocked = ~coverage & ~blocked
    if not np.any(uncovered_and_not_blocked):
        return word_grid

    uncovered_and_not_blocked_with_letter = (
        word_grid != GridState.OPEN
    ) & uncovered_and_not_blocked

    if np.any(uncovered_and_not_blocked_with_letter):
        to_fill = uncovered_and_not_blocked_with_letter
    else:
        to_fill = uncovered_and_not_blocked

    # Otherwise, we need to pick some of our to_fill spots for us to iterate over and check. Each of these are candidates for our choice to fill words from.
    candidate_locations = list(zip(*np.where(to_fill)))
    np.random.shuffle(candidate_locations)
    candidate_locations = candidate_locations[:max_checked_locations]

    # Prepare the scoring function with up to date information about the current grid state
    open_slots = word_grid == GridState.OPEN

    def _score(
        word_indices: Tuple[List[int], List[int]],
        word_rank: float,
        random_value: float,
    ) -> float:
        return score_word(coverage, open_slots, word_indices, word_rank, random_value)

    best_location_score = 0
    best_location_options = None

    for x, y in candidate_locations:
        current_options = []

        for slice_indices, slice, index_in_slice in all_matrix_slices_at(
            word_grid, x, y
        ):
            if len(slice) < trie.word_limits()[0] or np.all(slice != GridState.OPEN):
                continue

            current_options.extend(
                trie.get_top_k_valid_words(
                    slice=slice,
                    blocked=blocked[slice_indices],
                    indices=slice_indices,
                    target_index=index_in_slice,
                    scoring_function=_score,
                )
            )

        if len(current_options) == 0:
            # If we can't fill this spot with anything, then we need to backtrack, since nothing we do will change that.
            return None

        current_location_score = sum([score for score, _, _ in current_options])
        if (
            best_location_options is None
            or len(current_options) < len(best_location_options)
            or (
                len(current_options) == len(best_location_options)
                and current_location_score > best_location_score
            )
        ):
            best_location_options = current_options
            best_location_score = current_location_score

    # Now we know that there must be a best location

    # Now that we have found the most promising position, we will order our
    # words by their scores, and try each one at a time, moving to the next if
    # we end up backtracking.
    best_location_options.sort(key=lambda x: x[0], reverse=True)
    for _, word, indices in best_location_options:
        # If this option would not change our current board, then we don't need to place it, since that won't progress our solution.
        if np.sum(~coverage[indices]) == 0:
            continue

        # Place the word
        new_word_grid = word_grid.copy()
        new_word_grid[indices] = word
        new_coverage = grid_coverage(new_word_grid, trie)

        if not at_least_one_option_per_non_blocked_spot(
            word_search_grid=new_word_grid,
            blocked=blocked,
            coverage=new_coverage,
            trie=trie,
        ):
            # There is an unblocked, uncovered location which cannot be filled
            continue

        if np.any(new_coverage & blocked):
            # There is a blocked letter which was covered accidentally
            # This can happen if the blocked letter is an 's', for example, and
            # we put 'cat' right next to it, to make 'cats'.
            continue

        backtracking_result = fill_grid_backtracking(
            word_grid=new_word_grid,
            coverage=new_coverage,
            trie=trie,
            progress_bar=progress_bar,
            max_checked_locations=max_checked_locations,
        )
        if backtracking_result is not None:
            return backtracking_result

    return None


def seed_grid(word_grid: np.ndarray, density: float = 0.05) -> np.ndarray | None:  #
    width, height = word_grid.shape
    current_fill = np.sum(word_grid != GridState.OPEN)
    num_samples = int((word_grid.size - current_fill) * density)
    rx, ry = np.random.randint(0, width, num_samples * 5), np.random.randint(
        0, height, num_samples * 5
    )
    rv = np.random.randint(0, 26, num_samples * 5)

    new_word_grid = word_grid.copy()
    num_placed = 0
    for x, y, v in zip(rx, ry, rv):
        if new_word_grid[x, y] == GridState.OPEN:
            new_word_grid[x, y] = v
            num_placed += 1
            if num_placed >= num_samples:
                return new_word_grid

    return None


def solve(
    word_grid: np.ndarray,
    blocked: np.ndarray,
    trie: WordTrie,
    progress_bar: bool = True,
) -> WordSearch | None:
    if progress_bar:
        progress_bar = tqdm(total=word_grid.size)

    seeded_grid = seed_grid(word_grid)
    print(convert_matrix_to_letters(seeded_grid))
    if progress_bar:
        num_filled = np.sum(seeded_grid != GridState.OPEN)
        progress_bar.n = num_filled
        progress_bar.refresh()

    initial_coverage = grid_coverage(seeded_grid, trie)

    # Ensure that none of our blocked entries have been covered
    if np.any(initial_coverage & blocked):
        return None

    result = fill_grid_backtracking(
        word_grid=seeded_grid,
        coverage=initial_coverage,
        trie=trie,
        progress_bar=progress_bar,
    )
    if result is None:
        return None

    if progress_bar:
        progress_bar.n = result.size
        progress_bar.refresh()

    final_coverage = grid_coverage(result, trie)
    final_coverage = final_coverage & blocked
    if not np.all(final_coverage):
        raise RuntimeError("Should not be possible!")  # TODO: better error

    words = grid_words(result, trie)
    return WordSearch(result, words)


if __name__ == "__main__":
    from util import string_to_alphabet_positions

    # 0-25 Letters
    # 26 Unset / Needs to be solved for
    # np.inf Not a valid location

    # TODO: the way we are doing this right now means that we don't backtrack up through the open_slots cover stuff
    # when we hit a snag and need to backtrack

    trie = WordTrie()
    print(f"Constructed trie of depth {trie.depth()} from {len(trie)} words...")

    # grid = np.array(
    #     [
    #         [26, 8, 26, 26, 26, 11],
    #         [26, 26, 14, 26, 26, 26],
    #         [21, 26, 26, 26, 4, 26],
    #         [26, 26, 26, 26, 26, 26],
    #         [26, 26, 24, 26, 26, 14],
    #         [26, 26, 20, 26, 26, 26],
    #     ],
    #     dtype=int,
    # )
    grid = np.ones((6, 6), dtype=int) * 26
    message = "te amo"
    message = message.replace(" ", "").lower().strip()
    message_integers = string_to_alphabet_positions(message)
    print(message_integers)
    blocked = np.zeros_like(grid)
    locations = [(0, 0), (4, 1), (4, 5), (1, 2), (0, 4)]
    locations.sort()
    for location, value in zip(locations, message_integers):
        blocked[location] = 1
        grid[location] = value
    print(convert_matrix_to_letters(grid))
    print(blocked)
    print(solve(grid, blocked, trie))

File: src/generate.rs
use std::collections::HashSet;

use crate::consts::{OPEN, WORD_SEARCH_HEIGHT, WORD_SEARCH_SIZE, WORD_SEARCH_WIDTH};
use crate::grid::Grid;
use crate::trie::TrieNode;
use crate::util;

pub struct WordSearch<const N: usize, const ROWS: usize, const COLS: usize> {
    pub word_grid: Grid<char, N, ROWS, COLS>,
    pub words: HashSet<String>,
}

impl<const N: usize, const ROWS: usize, const COLS: usize> WordSearch<N, ROWS, COLS> {
    pub fn new(
        word_grid: Grid<usize, N, ROWS, COLS>,
        words: HashSet<String>,
    ) -> Result<Self, String> {
        let word_grid = word_grid.try_map(util::int_to_char)?;
        Ok(WordSearch {
            word_grid: word_grid,
            words: words,
        })
    }
}

fn score_word(
    coverage: Vec<bool>,
    open_slots: Vec<bool>,
    word_indices: (Vec<usize>, Vec<usize>),
    word_rank: f32,
    random_value: f32,
) -> f32 {
    0.0
}

fn grid_coverage<const N: usize, const ROWS: usize, const COLS: usize>(
    word_grid: &Grid<usize, N, ROWS, COLS>,
    trie: &TrieNode,
) -> Grid<bool, N, ROWS, COLS> {
    let mut coverage_grid = Grid::<bool, N, ROWS, COLS>::new();

    for slice in Grid::<usize, N, ROWS, COLS>::slices() {
        let (rows, cols) = slice;
        let data = word_grid
            .get_bulk(rows, cols)
            .copied()
            .collect::<Vec<usize>>();
        let slice_coverage = trie.slice_coverage(&data);
        let new_coverage = coverage_grid
            .get_bulk(rows, cols)
            .zip(&slice_coverage)
            .map(|(current, next)| *current || *next)
            .collect::<Vec<bool>>();
        coverage_grid.set_bulk(rows, cols, &new_coverage);
    }

    coverage_grid
}

fn grid_words<const N: usize, const ROWS: usize, const COLS: usize>(
    word_grid: &Grid<usize, N, ROWS, COLS>,
    trie: &TrieNode,
) -> HashSet<String> {
    let mut words = HashSet::new();

    for slice in Grid::<usize, N, ROWS, COLS>::slices() {
        let (rows, cols) = slice;
        let data = word_grid
            .get_bulk(rows, cols)
            .copied()
            .collect::<Vec<usize>>();
        let slice_words = trie.slice_words(&data);
        words.extend(slice_words);
    }

    words
}

fn possible_to_fill<const N: usize, const ROWS: usize, const COLS: usize>(
    word_grid: &Grid<usize, N, ROWS, COLS>,
    blocked: &Grid<bool, N, ROWS, COLS>,
    trie: &TrieNode,
    x: &usize,
    y: &usize,
) -> bool {
    for ((rows, cols), target_index) in Grid::<usize, N, ROWS, COLS>::slices_at(x, y) {
        let slice = word_grid.get_bulk(rows, cols).copied().collect::<Vec<_>>();
        let blocked_slice = blocked.get_bulk(rows, cols).copied().collect::<Vec<_>>();

        for _ in trie
            .get_valid_words(&slice, &blocked_slice, target_index)
            .enumerate()
        {
            return true;
        }
    }

    return false;
}

fn fill_grid_backtracking<const N: usize, const ROWS: usize, const COLS: usize>(
    word_grid: &Grid<usize, N, ROWS, COLS>,
    coverage: &Grid<bool, N, ROWS, COLS>,
    blocked: &Grid<bool, N, ROWS, COLS>,
    trie: &TrieNode,
) -> Option<Grid<usize, N, ROWS, COLS>> {
    let not_covered_and_not_blocked = coverage.not().and(&blocked.not());

    if !not_covered_and_not_blocked.any() {
        return Some(word_grid.clone());
    }

    let not_covered_and_not_blocked_with_letter = word_grid
        .map(|value| *value != OPEN)
        .and(&not_covered_and_not_blocked);

    // First, figure out which locations we will be filling
    let candidate_locations = if not_covered_and_not_blocked_with_letter.any() {
        not_covered_and_not_blocked_with_letter
            .items()
            .filter_map(|((row, col), is_not_covered_and_not_blocked)| {
                if *is_not_covered_and_not_blocked {
                    Some((row, col))
                } else {
                    None
                }
            })
            .collect::<Vec<_>>()
    } else {
        not_covered_and_not_blocked
            .items()
            .filter_map(|((row, col), is_not_covered_and_not_blocked)| {
                if *is_not_covered_and_not_blocked {
                    Some((row, col))
                } else {
                    None
                }
            })
            .collect::<Vec<_>>()
    };

    // Verify that each spot has at least one option
    for (x, y) in &candidate_locations {
        if !possible_to_fill(word_grid, blocked, trie, x, y) {
            return None;
        }
    }

    // Now consider each candidate location to find the most restricted one
    // let open_slots = word_grid.map(|e| *e == OPEN);

    const K: usize = 1000;
    let mut best_location_options = Vec::with_capacity(K * 4);
    let mut have_best = false;

    for (x, y) in &candidate_locations {
        let mut options_at_location = Vec::with_capacity(K * 4);

        for ((rows, cols), target_index) in Grid::<usize, N, ROWS, COLS>::slices_at(x, y) {
            let slice = word_grid.get_bulk(rows, cols).copied().collect::<Vec<_>>();
            let blocked_slice = blocked.get_bulk(rows, cols).copied().collect::<Vec<_>>();

            for (idx, (word_info, (start, stop))) in trie
                .get_valid_words(&slice, &blocked_slice, target_index)
                .enumerate()
            {
                if idx >= K {
                    break;
                }
                let (word_rows, word_cols) = (&rows[start..stop], &cols[start..stop]);
                options_at_location.push((0.0, word_info.path.clone(), (word_rows, word_cols)));
            }
        }

        if have_best {
            if options_at_location.len() < best_location_options.len() {
                best_location_options = options_at_location;
            }
        } else {
            best_location_options = options_at_location;
            have_best = true;
        }
    }

    // Now that we have a best location, we can iterate through our options there
    for (_, word, (rows, cols)) in best_location_options {
        // Prune branches that add words which are already in the grid
        if coverage
            .get_bulk(rows, cols)
            .fold(0, |acc, e| if *e { acc } else { acc + 1 })
            == 0
        {
            continue;
        }

        let mut hypothetical_grid = word_grid.clone();
        hypothetical_grid.set_bulk(rows, cols, &word);
        let hypothetical_coverage = grid_coverage(&hypothetical_grid, trie);

        // Make sure that we haven't covered a blocked letter by accident. (If the blocked letter is an s, for example)
        if hypothetical_coverage.and(blocked).any() {
            continue;
        }

        match fill_grid_backtracking(&hypothetical_grid, &hypothetical_coverage, blocked, trie) {
            Some(grid) => return Some(grid),
            None => {}
        }
    }

    None
}

fn seed_grid<const N: usize, const ROWS: usize, const COLS: usize>(
    word_grid: Grid<usize, N, ROWS, COLS>,
    density: f32,
) -> Grid<usize, N, ROWS, COLS> {
    let unfilled_indices = word_grid
        .items()
        .filter_map(|((row, col), value)| {
            if *value == OPEN {
                Some((row, col))
            } else {
                None
            }
        })
        .collect::<Vec<_>>();
    let target_num_filled = (N as f32 * density).round() as i32;
    let num_to_fill = target_num_filled - (N as i32 - unfilled_indices.len() as i32);

    if num_to_fill < 0 {
        return word_grid;
    }

    let num_to_fill = num_to_fill as usize;
    let mut new_grid = word_grid.clone();

    // TODO: generate `num_to_fill` random integers in the range 0..unfilled_indices.len() without replacement
    // Then select those integers from unfilled_indices and give them random values

    word_grid
}

pub fn solve<const N: usize, const ROWS: usize, const COLS: usize>(
    word_grid: Grid<usize, N, ROWS, COLS>,
    blocked: Grid<bool, N, ROWS, COLS>,
    trie: TrieNode,
) -> Option<WordSearch<N, ROWS, COLS>> {
    let seeded_grid = seed_grid(word_grid, 0.05);
    let initial_coverage = grid_coverage(&seeded_grid, &trie);

    // Make sure we haven't blocked anything
    for (coverage, blocked) in initial_coverage.values().zip(blocked.values()) {
        if *coverage && *blocked {
            return None;
        }
    }

    let result = fill_grid_backtracking(&seeded_grid, &initial_coverage, &blocked, &trie);

    match result {
        None => None,
        Some(solved_grid) => {
            let final_coverage = grid_coverage(&solved_grid, &trie);

            // Make sure we solved correctly
            for (is_covered, is_blocked) in final_coverage.values().zip(blocked.values()) {
                if !*is_covered && !*is_blocked {
                    panic!()
                }
            }

            let words = grid_words(&solved_grid, &trie);

            match WordSearch::new(solved_grid, words) {
                Err(_) => None,
                Ok(word_search) => Some(word_search),
            }
        }
    }
}
